test/stress/s3benchmark.cc

```diff
@@ -0,0 +1,266 @@
+  upload::S3Uploader *uploader_;
+  vector<TestDataChunk> data_chunks_;
+  int unique_files_ = 0;  // counter for unique files uploaded
+  Prng prng_;
+  int next_hash_num_ = 1;
+
+  void GeneratePayload();
+  TestDataChunk GenerateChunk();
+  TestDataChunk DuplicateChunk(const TestDataChunk &old_chunk);
+  void UploadFile(const TestDataChunk &chunk);
+  float RandomNormal();
+  double CalculateDuration(const timespec &start, const timespec &end);
+};
+
+S3TestScenario::S3TestScenario(
+  const string &config_path,
+  const string &temp_path,
+  int num_uploads,
+  int avg_file_size,
+  float duplicate_file_ratio)
+  : config_path_(config_path)
+  , tmp_path_(temp_path)
+  , num_uploads_(num_uploads)
+  , avg_file_size_(avg_file_size)
+  , duplicate_file_ratio_(duplicate_file_ratio)
+{
+  prng_.InitLocaltime();
+
+  assert(MkdirDeep(tmp_path_, 0755, true));
+
+  string definition_string = "S3," + tmp_path_ +
+                             "/,s3benchmark_cvmfs@" + config_path;
+  spooler_definition_ = new upload::SpoolerDefinition(
+    definition_string,
+    shash::Algorithms::kSha1);
+  uploader_ = new upload::S3Uploader(*spooler_definition_);
+}
+
+int S3TestScenario::Run()
+{
+  uint64_t start, payload_generated, files_uploaded,
+           files_reuploaded, files_deleted;
+  uploader_->Initialize();
+  start = platform_monotonic_time_ns();
+  GeneratePayload();
+  payload_generated = platform_monotonic_time_ns();
+  for (const TestDataChunk &chunk : data_chunks_) {
+    UploadFile(chunk);
+  }
+  uploader_->WaitForUpload();
+  files_uploaded = platform_monotonic_time_ns();
+  for (const TestDataChunk &chunk : data_chunks_) {
+    UploadFile(chunk);
+  }
+  uploader_->WaitForUpload();
+  files_reuploaded = platform_monotonic_time_ns();
+  for (const TestDataChunk &chunk : data_chunks_) {
+    string path = "data/" + chunk.hash.MakePath();
+    uploader_->RemoveAsync(path);
+  }
+  uploader_->WaitForUpload();
+  files_deleted = platform_monotonic_time_ns();
+  double duration_payload = (payload_generated - start)*1e-9;
+  double duration_upload = (files_uploaded - payload_generated)*1e-9;
+  double duration_reupload = (files_reuploaded - files_uploaded)*1e-9;
+  double duration_delete = (files_deleted - files_reuploaded)*1e-9;
+
+  LogCvmfs(kLogCvmfs, kLogStdout, "Finished %d uploads, %d unique files "
+                       "of average size %d bytes\n"
+                       "Payload generation: %f seconds\n"
+                       "HEAD(NotFound)+PUT(New): %f seconds\n"
+                       "HEAD(Found)+PUT(Existing): %f seconds\n"
+                       "DELETE: %f seconds", num_uploads_, unique_files_,
+                       avg_file_size_, duration_payload, duration_upload,
+                       duration_reupload, duration_delete);
+
+  return 0;
+}
+
+void S3TestScenario::GeneratePayload() {
+  for (int i = 0; i < num_uploads_; ++i) {
+    float duplicate_chance = prng_.Next(INT_MAX) * (1.0 / INT_MAX);
```

Comment:
I think we would like to have that part of the `Prng` class, like `NextP` or `NextProbability`